astropy/io/fits/hdu/compressed/header.py

```diff
@@ -1008,16 +769,37 @@ def _image_header_to_bintable_header_and_coldefs(
     # the image header.  This allows those blank cards to be carried
     # over to the image header when the hdu is uncompressed.
 
-    if "ZHECKSUM" in bintable_header:
+    if "ZHECKSUM" in bintable.header:
         required_blanks = image_header._countblanks()
-        image_blanks = generated_image_header._countblanks()
-        table_blanks = bintable_header._countblanks()
+        image_blanks = image_header._countblanks()
+        table_blanks = bintable.header._countblanks()
 
         for _ in range(required_blanks - image_blanks):
-            generated_image_header.append()
+            image_header.append()
             table_blanks += 1
 
         for _ in range(required_blanks - table_blanks):
-            bintable_header.append()
+            bintable.header.append()
+
+    bintable.columns = cols
+
+    # Add any keywords that are in the original header that are not already
+    # FIXME: don't use keyword_remaps, instead define an actual list to check
+    # including regular expressions for NAXIS and other similar keywords
+    for card in image_header.cards:
+        if card.keyword == "":
+            continue  # TODO: investigate why some cards have empty keyword
+        elif card.keyword == "COMMENT":
+            bintable.header.add_comment(card.value)
+        elif card.keyword == "HISTORY":
+            bintable.header.add_history(card.value)
+        elif (
+            card.keyword not in CompImageHeader._keyword_remaps
+            and card.keyword not in bintable.header
+            and not card.keyword.startswith("NAXIS")
+        ):
+            bintable.header.append(card)
```

Comment:
Could you clarify what you mean here? I think the loop will skip over blank cards currently.

Reply:
Yes but `.append` replaces blank cards by default (`useblanks=True`). So not sure but it's possible that the change in this loop has some side effects, because it is now adding at the same time blanks and other cards while previously blanks were added at the end.